src/Perpetuum/Zones/Effects/ZoneEffects/ZoneEffectHandler.cs

```diff
@@ -1,22 +1,86 @@
-﻿using Perpetuum.Players;
+﻿using Perpetuum.Log;
+using Perpetuum.Players;
+using Perpetuum.Units;
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
 
 namespace Perpetuum.Zones.Effects.ZoneEffects
 {
     public class ZoneEffectHandler : IZoneEffectHandler
     {
-        private readonly IZoneEffectRepository _zoneEffectRepository;
-        public ZoneEffectHandler(IZoneEffectRepository zoneEffectRepository)
+        private const byte _ = 0;
+        private readonly IZone _zone;
+        private readonly ConcurrentDictionary<ZoneEffect, byte> _effects;
+
+        public ZoneEffectHandler(IZone zone)
+        {
+            _zone = zone;
+            _effects = new ConcurrentDictionary<ZoneEffect, byte>();
+            InitCollection();
+        }
+
+        private void InitCollection()
+        {
+            foreach (var zoneEffect in ZoneEffectReader.GetStaticZoneEffects(_zone))
+            {
+                _effects.Add(zoneEffect, _);
+            }
+        }
+
+        public void AddEffect(ZoneEffect zoneEffect)
+        {
+            _effects.TryAdd(zoneEffect, _);
+            OnZoneEffectAdded(zoneEffect);
+        }
+
+        public void RemoveEffect(ZoneEffect zoneEffect)
+        {
+            _effects.TryRemove(zoneEffect, out byte b);
+            OnZoneEffectRemoved(zoneEffect);
+        }
+
+        private void OnZoneEffectRemoved(ZoneEffect effect)
+        {
+            var eligibleUnits = _zone.Units.Where(u => u.EffectHandler.ContainsEffect(effect.Effect));
+            foreach (var unit in eligibleUnits)
+            {
+                unit.EffectHandler.RemoveEffectsByType(effect.Effect);
+                Logger.DebugInfo($"Removing {effect.Effect} to {unit} on zone:{_zone.Id}");
+            }
+        }
+
+        private void OnZoneEffectAdded(ZoneEffect effect)
+        {
+            var eligibleUnits = _zone.Units.Where(u => CanApplyEffect(effect)(u));
+            foreach (var unit in eligibleUnits)
+            {
+                var builder = unit.NewEffectBuilder().SetType(effect.Effect).SetOwnerToSource();
+                unit.ApplyEffect(builder);
+                Logger.DebugInfo($"Adding {effect.Effect} to {unit} on zone:{_zone.Id}");
+            }
+        }
+
+        private IEnumerable<ZoneEffect> GetEffects()
+        {
+            return _effects.Keys;
+        }
+
+        private Func<Unit, bool> CanApplyEffect(ZoneEffect effect)
         {
-            _zoneEffectRepository = zoneEffectRepository;
+            return u => !u.EffectHandler.ContainsEffect(effect.Effect) && effect.PlayerOnly == u is Player;
```

Comment:
This may have an unintended side effect. If `effect.PlayerOnly` is false (allowing it to affect everything) and `u` is a `Player` then the statement will fail even though it shouldn't.

Reply:
Oh shoot you are right, I was not careful with the false/true case which should still allow effect application